### ISSUE TITLE
A Polling Operation

### ISSUE THREAD

**Issue by jshier** [danthorpe/Operations#153]
My current project has several network endpoints that I'm required to poll against. This is due to backend design, not desire, and I have no control over this requirement. As the number of endpoints I need to poll against has increased, my need for a generic solution has as well. So I've started doing so and while I have a prototype working, additional guidance would be appreciated. Here's what it looks like so far:
```swift
import Operations

typealias OperationGenerator = () -> Operation

class PollingOperation : GroupOperation {
    var activeOperation: Operation
    var delayOperation: DelayOperation
    
    private let operationGenerator: OperationGenerator
    private let interval: NSTimeInterval
    
    init(interval: NSTimeInterval, operationGenerator: OperationGenerator) {
        self.interval = interval
        self.operationGenerator = operationGenerator
        activeOperation = operationGenerator()
        delayOperation = DelayOperation(interval: interval)
        super.init(operations: [])
    }
    
    override func execute() {
        continuePolling()
        
        super.execute()
    }
    
    override func operationDidFinish(operation: NSOperation, withErrors errors: [ErrorType]) {
        if let pollable = operation as? Pollable {
            if pollable.shouldContinuePolling() {
                continuePolling()
            }
            else {
                finish()
            }
        }
    }
    
    func continuePolling() {
        delayOperation = DelayOperation(interval: interval)
        activeOperation = operationGenerator()
        activeOperation.addDependency(delayOperation)
        
        addOperations([delayOperation, activeOperation])
    }
}

protocol Pollable {
    func shouldContinuePolling() -> Bool
}
```
My main issue at the moment is how to make the `Pollable` protocol fully generic so I can add conformance to any arbitrary operation. Here's the current `TestOperation`:
```swift
class TestOperation : Operation {
    
    override init() {
        super.init()
        name = "Test Operation"
    }
    
    override func execute() {
        print("*** Executing TestOperation. ***")
        finish()
    }
}

extension TestOperation : Pollable {
    func shouldContinuePolling() -> Bool {
        let timeInterval = NSDate().timeIntervalSince1970
        let remainder = Int64(timeInterval) % 10
        
        print("*** shouldContinuePolling() for interval: \(timeInterval), remainder: \(remainder). ***")
        return (remainder != 0)
    }
}
```
Called like this:
```swift
let pollingOperation = PollingOperation(interval: 1) { () -> Operation in
    return TestOperation()
}
let queue = OperationQueue()
queue.addOperation(pollingOperation)
```

This works well in my limited testing. But I'd like to require `OperationGenerator` to create `Pollable` `Operation`s but I can't seem to express that requirement in Swift. The initializer could use some work as well, as I don't like generating unnecessary operations just to get past the Swift initializer requirements.

In general, there may also be a use for an even more generic `RepeatableOperation` protocol. I'd appreciate any guidance or ideas here.

**Comment by danthorpe** [comment]
This is really interesting, as I've been thinking about creating repeating / retrying / exponential back-off `Operation` subclasses for v2.6.

My gut reaction to this is that you could make it a generic operation:

```swift
class PollingOperation<O where O: NSOperation, O: Pollable>: GroupOperation {
   let creator: () -> O
   let timeInterval: NSTimeInterval

   init(timeInterval: NSTimeInterval, creator: () -> O) {
     // etc
   }
}
```

I don't think there is a way to do it by using a typealias for the closure outside of a generic scope.

Take a look at [`ReachableOperation`](https://github.com/danthorpe/Operations/blob/development/Sources/Features/Shared/ReachableOperation.swift#L59) for a similar kind of repeating architecture too.

**Comment by danthorpe** [comment]
@jshier I've done some work on a `RepeatedOperation` in #154 - not yet merged. But would be great if you have any feedback on this.

Using the example here, your use case would work like this:

```swift

extension TestOperation: Repeatable {
    func shouldRepeat(count: Int) -> Bool {
        let timeInterval = NSDate().timeIntervalSince1970
        let remainder = Int64(timeInterval) % 10

        print("*** shouldContinuePolling() for interval: \(timeInterval), remainder: \(remainder). ***")
        return (remainder != 0)
    }
}

let operation = RepeatedOperation(strategy: .Fixed(1.0)) { TestOperation() }
queue.addOperation(operation)
```

In this case, the `strategy` argument is a fixed `NSTimeInterval` of 1 second. See the notes in #154 for more info.

The `shouldContinuePolling()` (which I've copied into the `Repeatable` extension)  is interesting... I'm not entirely sure what your intention here is, from reading the code, this seems to me that when the operation finishes then 9/10 times another operation is added, else the group finished polling. The 0.1 failure probability because you're effectively sampling a really fast clock for when the microseconds accuracy is 0. I'd have thought it's better to just use `arc4random` to determine this, but also wondering if this is what you really want?

Also - and this is probably a typo, in `GroupOperation` subclasses, calling `finish()` from within `operationDidFinish:withErrors:` is not necessary. The `GroupOperation` calls this function to allow subclasses to add more operations to the group, it has no other purpose. Internally the group has a "finishing operation" which is waiting on all the child operations and when this completes, the group operation finishes. Therefore, there is no need to explicitly call `finish` which will terminate the group prematurely. I've not got any tests for this, but it might cause some unexpected behavior.

**Comment by jshier** [comment]
Thanks, I'll take a look when I get a chance.

The `shouldContinuePolling()` function was supposed to be general so that the operation being repeated can determine whether polling should continue. The logic that's in there for the `TestOperation` was just to make sure I could easily poll for a dynamic number of times. The logic I actually need is to repeat a network request until the response changes in a way to indicate completion, at which point the polling should stop. I haven't looked at what you've done, but another requirement was to be able to stop polling after a certain number of a certain type of error occurred, which I hadn't yet figured out in the context of a abstract operation.

**Comment by jshier** [comment]
Also, I didn't realize that `finish()` shouldn't be called from `operationDidFinish`. I'll make sure that I audit my operations for that.

**Comment by danthorpe** [comment]
@jshier okay, I thought that's what you were wanting to do. I've added `RandomFailGenerator` to support this kind of random failure in general. Use like this:

```swift
let operation = RepeatedOperation(RandomFailGenerator(anyGenerator { TestOperation() }, probability: 0.01))
```

which randomly fails (i.e. finishes the repeating) ~ 1 out of every 100 operations.

So far, I think my `Repeatable` protocol is generic enough to support the current (actually just finished) instance from determining whether a new operation should be added. It receives the current count of executed operations. But I will also now include the aggregate errors from the group too. This would then allow that repeating operation to inspect the probability distribution of errors.

The `RepeatedOperation` also supports a "max count" for tries too.

**Comment by danthorpe** [comment]
Hmm, actually will need to think a little bit more about including the aggregate errors into `Repeatable`. At the moment, `RepeatedOperation` is nicely de-coupled from `Repeatable` - as it's only generic over the operation type, using `AnyGenerator` to erase the type of the generator. To pass the aggregate errors through to the generator would require a specific generator type.

**Comment by danthorpe** [comment]
Okay, so I'm not going to modify `RepeatedOperation` to support aggregating errors. Why? Because it tightly couples the concept of repeatedly executing operations to the logic for whether or not they should be repeated (using a particular metric).

Instead, that logic should go into the generator of the operations itself. 

However, to help with this, I've added `RepeatableOperation` in #154. It can be used to aggregate whatever properties you want and then use them to determine whether to repeat the operation or not. E.g. to stop repeating once a threshold of errors is reached:

```swift
var errors: [ErrorType] = []
let operation = RepeatedOperation { () -> RepeatableOperation<TestOperation> in

    let op = TestOperation(error: TestOperation.Error.SimulatedError)
    op.addObserver(FinishedObserver { _, e in
        errors.appendContentsOf(e)
    })

    return RepeatableOperation(op) { count in 
        return count < 10 && errors.count < 3 
    }
}
```

Basically `RepeatableOperation` is an `Operation` which composes another `Operation` and is initialized with a block. The block receives the current `count` of executed operations. We can add observers to the composed operation (or even just the `RepeatableOperation`) to aggregate state which can be checked inside the block.

It's perhaps not the most prettiest code ever - but it is the most flexible technique. Also it means that existing operations, which perhaps don't conform to `Repeatable` can be *made* to be repeatable.

**Comment by danthorpe** [comment]
Hi @jshier the `development` branch now includes `RepeatedOperation` and `RetryOperation` from #154 (in addition to 2.5.1). Which I think will support your need for a polling operation directly.

I'm going to close this issue for now, if you have feedback regarding `RepeatedOperation` - feel free to re-open this issue, or start a new one.
